Sources/FoundationEssentials/Calendar/Calendar_Recurrence.swift

```diff
@@ -0,0 +1,836 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Recurrence enumeration
+//
+// This file implements enumerating occurences according to a recurrence rule as
+// specified in RFC5545 and RFC7529
+
+@available(FoundationPreview 0.4, *)
+extension Calendar.RecurrenceRule.Frequency {
+    /// The calendar component corresponding to the frequency interval
+    var component: Calendar.Component {
+        switch self {
+        case .minutely: .minute
+        case .hourly:   .hour
+        case .daily:    .day
+        case .weekly:   .weekOfMonth
+        case .monthly:  .month
+        case .yearly:   .year
+        }
+    }
+}
+
+/// The action of a component of the recurrence rule.
+///
+/// Each field in the recurrence rule has a different effect on the output which
+/// is also different for different frequencies. The effect is documented in the
+/// RFC, and copied below for convenience:
+///
+/// | Component\Freq |.minutely|.hourly |.daily  |.weekly |.monthly|.yearly |
+/// |----------------|---------|--------|--------|--------|--------|--------|
+/// |months          |.limit   |.limit  |.limit  |.limit  |.limit  |.expand |
+/// |weeks           |nil      |nil     |nil     |nil     |nil     |.expand |
+/// |daysOfTheYear   |.limit   |.limit  |nil     |nil     |nil     |.expand |
+/// |daysOfTheMonth  |.limit   |.limit  |.limit  |nil     |.expand |.expand |
+/// |weekdays        |.limit   |.limit  |.limit  |.expand |note 1  |note 2  |
+/// |hours           |.limit   |.limit  |.expand |.expand |.expand |.expand |
+/// |minutes         |.limit   |.expand |.expand |.expand |.expand |.expand |
+/// |setPositions    |.limit   |.limit  |.limit  |.limit  |.limit  |.limit  |
+///
+/// - Note 1: Limit if `daysOfTheMonth` is present; otherwise expand
+/// - Note 2: Limit if `daysOfTheYear` or `daysOfTheMonth` is present; otherwise
+///           expand
+enum ComponentAction {
+    /// Replace each date found so far with a list of dates
+    case expand
+    /// Remove some of the dates which have been found so far
+    case limit
+}
+
+extension Calendar {
+    /// A `Sequence` of `Date`s produced by a given recurrence rule
+    @available(FoundationPreview 0.4, *)
+    struct DatesByRecurring : Sendable, Sequence {
```

Comment:
We could. Is there a precedent for other sequences being lazy by default? If we don't add conformance, the user can always use `.lazy` to lazily map and filter.

Reply:
There are some in other modules, but not in Swift-foundation. I'm not proposing for doing that here of course since that would require API proposal, just something we can consider in the future since we were talking about performance